searx/settings_loader.py

```diff
@@ -1,68 +1,97 @@
+DEFAULT_SETTINGS_PATH = Path(searx_dir) / SETTINGS_YAML
+"""Location of the default SETTINGS_YAML file."""
 
 
-def load_yaml(file_name):
+def load_yaml(file_name: str | Path):
+    """Load YAML config from a file."""
     try:
         with open(file_name, 'r', encoding='utf-8') as settings_yaml:
-            return yaml.safe_load(settings_yaml)
+            cfg = yaml.safe_load(settings_yaml)
+            if cfg is None:
+                cfg = {}
+            return cfg
     except IOError as e:
-        raise SearxSettingsException(e, file_name) from e
+        raise SearxSettingsException(e, str(file_name)) from e
     except yaml.YAMLError as e:
-        raise SearxSettingsException(e, file_name) from e
+        raise SearxSettingsException(e, str(file_name)) from e
+
+
+def get_yaml_cfg(file_name: str | Path) -> dict:
+    """Shortcut to load a YAML config from a file, located in the
+
+    - :py:obj:`get_user_cfg_folder` or
+    - in the ``searx`` folder of the SearXNG instalation
+    """
 
+    folder = get_user_cfg_folder()
+    if not folder:
+        folder = Path(searx_dir)
+    fname = folder / file_name
+    if not fname.is_file():
+        raise FileNotFoundError(f"File {fname} does not exist!")
 
-def get_yaml_file(file_name):
-    path = existing_filename_or_none(join(searx_dir, file_name))
-    if path is None:
-        raise FileNotFoundError(f"File {file_name} does not exist!")
+    return load_yaml(fname)
 
-    return load_yaml(path)
 
+def get_user_cfg_folder() -> Path | None:
+    """Returns folder where the user settings are located.
 
-def get_default_settings_path():
-    return existing_filename_or_none(join(searx_dir, 'settings.yml'))
+    1. If environment ``SEARXNG_SETTINGS_PATH`` is set
+       (e.g. ``/etc/mysxng/myinstance.yml``) then the path name for this
+       configuration file is the folder in which the user settings are expected
+       (e.g. ``/etc/mysxng``).  If this folder does not exists a
+       :py:obj:`EnvironmentError` is raised.
 
+    2. If folder ``/etc/searxng`` exists it is used.
 
-def get_user_settings_path() -> Optional[str]:
-    """Get an user settings file.
-    By descending priority:
-    1. ``environ['SEARXNG_SETTINGS_PATH']``
-    2. ``/etc/searxng/settings.yml`` except if ``SEARXNG_DISABLE_ETC_SETTINGS`` is ``true`` or ``1``
-    3. ``None``
+    In case none of the above path exists, ``None`` is returned.
     """
 
-    # check the environment variable SEARXNG_SETTINGS_PATH
-    # if the environment variable is defined, this is the last check
-    if 'SEARXNG_SETTINGS_PATH' in environ:
-        return existing_filename_or_none(environ['SEARXNG_SETTINGS_PATH'])
+    folder = Path("/etc/searxng")
+    settings_path = os.environ.get("SEARXNG_SETTINGS_PATH")
+    if settings_path:
+        folder = Path(settings_path).parent
+        if not folder.is_dir():
+            raise EnvironmentError(1, f"Folder of {settings_path} not exists!", folder)
```

Comment:
Shouldn't we first check if
1. is `SEARXNG_SETTINGS_PATH` a folder (e.g. `/etc/searxng`)? if yes, that's our settings path
2. we assume `SEARXNG_SETTINGS_PATH` is a file, thus we use its parent folder. if it exists, that's our settings path
3. we raise an exception
Currently, we only do step 2 and 3, but not step 1

Reply:
> Shouldn't we first check if … but not step 1

Good point I had already thought about it .. my first intention was not to change the meaning of `SEARXNG_SETTINGS_PATH`:

TL;DR;

Alternatively, I had thought about introducing a completely new environment variable for the config folder (SEARXNG_CFG_FOLDER).   Note that this settings loader is only for YAML files ... but we also have the toml file from the limiter .. and other plugins might bring more config file types ... all of them are in this folder.

But this would have the disadvantage that we have ambiguities (SEARXNG_CFG_FOLDER vs SEARXNG_SETTINGS_PATH).

Since I didn't want to choose one or the other with this PR, I decided not to change the meaning of SEARXNG_SETTINGS_PATH in this PR.

Reply:
I agree that adding a new option may cause confusion and `SEARXNG_SETTINGS_PATH` should continue to be used.

I just think that it might be more convenient to also support folder paths, now that there's the ability to use files other than `settings.yml` (e.g. `replace.yml` for the `hostnames plugin`) it's really rather a folder path than a single file path. It wouldn't really change the behavior to implement 1. (the old full settings file path still works), it's just additional convenience for instance admins (in my opinion).

Reply:
> I just think that it might be more convenient to also support folder paths,

Actually, you are right and my doubts are wrong ... I have now implemented it, see [diff](https://github.com/searxng/searxng/compare/c7d49e873b6a9ad787f0fcf59679d900934fa04f..1997d53c21bcd16ab70cd930aaa3a3c46394d674) .. further I changed the downward compatibility slightly; read my https://github.com/searxng/searxng/pull/3578#discussion_r1642433145